File: JavaScript/0145. Binary Tree Postorder Traversal.js
// Given a binary tree, return the postorder traversal of its nodes' res.
//
// Example:
//
// Input: [1,null,2,3]
//    1
//     \
//      2
//     /
//    3
//
// Output: [3,2,1]
// Follow up: Recursive solution is trivial, could you do it iteratively?

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

// 1) Recursion
const postorderTraversal1 = (root) => {
  const res = [];

  const go = (node) => {
    if (node == null) return;
    go(node.left);
    go(node.right);
    res.push(node.val);
  };

  go(root);
  return res;
};

// 2) Iteration using stack
const postorderTraversal2 = (root) => {
  const st = [];
  const res = [];

  if (root != null) st.push(root);

  while (st.length) {
    root = st.pop();
    res.unshift(root.val);
    if (root.left) st.push(root.left);
    if (root.right) st.push(root.right);
  }
  return res;
};

File: Python/0145. Binary Tree Postorder Traversal.py
# Given the root of a binary tree, return the postorder traversal of its nodes' values.
#
# Example 1:
#
# Input: root = [1,null,2,3]
# Output: [3,2,1]
#
# Example 2:
#
# Input: root = []
# Output: []
#
# Example 3:
#
# Input: root = [1]
# Output: [1]
#
# Constraints:
#
# The number of the nodes in the tree is in the range [0, 100].
# -100 <= Node.val <= 100


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


# 1) Recursion
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        res = []

        def go(node):
            if not node:
                return
            go(node.left)
            go(node.right)
            res.append(node.val)

        go(root)
        return res


class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        return (
            self.postorderTraversal(root.left)
            + self.postorderTraversal(root.right)
            + [root.val]
        )


# 2) Iteration
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        res = []
        st = [root]

        while st:
            node = st.pop()
            res.insert(0, node.val)
            if node.left:
                st.append(node.left)
            if node.right:
                st.append(node.right)
        return res
